### ISSUE TITLE
Commutative operator overloading

### ISSUE THREAD

**Issue by ChayimFriedman2** [wren-lang/wren#793]
When I first saw #746, I thought about Python.

In short, that issue talks about commutative operator overloading; that is, the ability to overload not just `<myclass> + <otherclass>` but also `<otherclass + myclass>`.
This is intuitive to anyone for numerous operators: `+`, `*`, etc. on numeric types and more (e.g. if I would want to implement a complex number class in Wren).

Unfortunately, the strategy used in Wren for operator overloading is simple: check if the left operand has a method named same as the operator, and call it with the right operand. This makes it impossible to overload operators commutatively.

Python solved this in two steps:

1. Define a constant `NotImplemented`, which any overloaded binary operator should return when there is type mismatch; in the example of the above issue, `Num` does not know how to multiply itself with `List`, so it'll return `NotImplemented`.
2. Define a reverse operator, `__r<operator>__`, for each binary operator. If the method of the left operand returned `NotImplemented` _and_ there is a reverse operator method on the right operand call it with the left operand as parameter. If it returned `NotImplemented` too, raise an error.

Wren can implement a similar strategy too. We can define a static method `Object.notImplemented`. About the second step, I have two ideas:

1. Just like Python, add reverse operator semantics. Something like a `reverse` keyword, e.g. `reverse *(other)`. This has the disadvantage of adding a keyword and requiring another method for the same operator (usually it'll just `return this * other`).
2. Define a specific set of operators as commutative, and automatically call them as reverse operators if needed. I do not like this approach, because although most of the time there's a constant set of commutative operators, there can still be cases where I want to overload both the reverse and the straight version of the operator. For example, if I'll continue with the complex numbers example from above, I'll want to be able to do both `complex / irrational` and `irrational / complex`, even though they have different meanings.

**Comment by mhermier** [comment]
This is a huge can of worm, with no real easy solution. The solution you
propose is valid, and try to solve the not having control of the source.
Another solution to this problem could be to allow to expand classes
outside of primary class definition, or granting an operator to be
commutative on some class.

**Comment by ChayimFriedman2** [comment]
@mhermier I liked the second solution, but I think the first breaks the nature of Wren. It'll also make it much less performant.

**Comment by mhermier** [comment]
Elaborate how it can be slow?
If you craft an opcode to swap arguments on the stack, and call the
operator again (and given it can be re-used per operator, I don't see how
it can be a penalty compared to do a full round.
First option implies, trying to do a call, asking querying the class, and
then perform the same evaluation as on previous definition.

**Comment by ChayimFriedman2** [comment]
I don't understand. Do you mean the following?

```wren
class Complex {
  construct fromNum() { /* ... */ }

  *(other) {
    if (other is Num) {
      return this * Complex.fromNum(other)
    }
    // ...
  }
  multiplyNum_(other) {
    // `this` is `Num`
    if (other is Complex) {
      return other * this
    }
    return Num.originalNumMultiply_(other)
}
Num.originalNumMultiply_ = Num.*
Num.* = Complex.multiplyNum_
```

**Comment by mhermier** [comment]
In essence yes, maybe not exactly that syntax. Or something like this:
```
class Complex {
}

Complex.* = Fn.new {|lhs rhs|...}
```
But details have to be thought, because this involves overload name
resolution which is another can of worms that need to be think of...

**Comment by ChayimFriedman2** [comment]
I thought about block functions, but rejected it because they requires `.call()`.

Except the boilerplate (which is huge, as you can see in my example) and the compromise of readability (again), this will make the VM slower because the speed of Wren comes mostly from the speed of method/fields location - they're known at compile-time. If we'll be able to change existing class structure (like other dynamic languages), we'll loss this advantage. Field and method access will need to perform a lookup in a dictionary, and it'll slower the VM in factors (I can guess more than 2x).

**Comment by mhermier** [comment]
Behind the scene it is a linear search, so it is only a matter of 
expanding an existing class, so speeds as this is not really an issue 
(as in current implementation). The only real issue is control where we 
allow expanding structs, so we don't get some oddities/corruption.

About the `.call()` issue, the argument is not a real issue. Since 
methods/foreign are not first class object (yet), it can state that they 
follow `.call()` *protocol*

**Comment by ChayimFriedman2** [comment]
> Behind the scene it is a linear search, so it is only a matter of expanding an existing class, so speeds as this is not really an issue (as in current implementation).

Linear search? What do you mean?

> The only real issue is control where we allow expanding structs, so we don't get some oddities/corruption.

structs in Wren?

> About the `.call()` issue, the argument is not a real issue. Since methods/foreign are not first class object (yet), it can state that they follow `.call()` *protocol*

If they **were** first-class they could implement a protocol, but now when they're not?!

Can you not put too many newlines please? It is hard to read...

**Comment by mhermier** [comment]
Classes holds methods in a linear and predictable way in the current 
implementation. Meaning if you define a method name `foo()`, it will get 
an index assigned, that will be available forever and every class during 
the lifetime of the VM. So expanding an old class to provide `foo()`, is 
only expanding the array of methods array so it can contains the index 
for `foo()` and assign the method at it.

**Comment by ChayimFriedman2** [comment]
Yes. But the index is provided at compile time, and with that it'll have to be generated at run time. See:
```wren
class A {
  f() {
    System.print(this.f2()) // `this` is required, or the compiler won't know where `f2()` is defined
  }
}
A.f2 = A.f // Of course infinite recursion, just for demonstration
```
When compiling, the compiler won't know how to access `A.f2()`, so it'll need to leave it as-is. And the interpreter will have to perform a string lookup.

**Comment by ChayimFriedman2** [comment]
And also this will be asymmetric and confusing, since it won't allow getters/setters/static functions.

Also the compiler will need to accept operators as class methods `Class.*`. Minor change, though.

**Comment by mhermier** [comment]
There is no such error, because the compiler is dumb. Whenever it finds 
and unknown methods, it creates an entry. So as in current state, it 
should be valid given method assignment works.

**Comment by ChayimFriedman2** [comment]
Also, now `this` outside of a method is a compile-time error. It will have to be a run-time error.

**Comment by ChayimFriedman2** [comment]
> There is no such error, because the compiler is dumb. Whenever it finds and unknown methods, it creates an entry. So as in current state, it should be valid given method assignment works.

OK, although that seems to be a bug in the compiler (accessing an undefined method will not raise an error).

**Comment by mhermier** [comment]
This starts to be really be out of control, adding too much work around. 
It make the function approach less intrusive.

**Comment by ChayimFriedman2** [comment]
> There is no such error, because the compiler is dumb. Whenever it finds and unknown methods, it creates an entry. So as in current state, it should be valid given method assignment works.

But there is still problem - consider:
```wren
class C {
}
class Helper {
  f1() {
    this.f2()
  }
  f2() {
  }
}
C.f1 = Helper.f1
C.f2 = Helper.f2
```

**Comment by mhermier** [comment]
It would be a compiler error, if a more advance type checking was in 
place. Currently type checking is in-existent at compile time, and is 
only done at runtime.

**Comment by ChayimFriedman2** [comment]
> This starts to be really be out of control, adding too much work around.

I absolutely agree. As I stated above, this is a complete change to the design of the language. Massive changes to the language requires massive changes to the implementation, naturally. This is the reason I rejected this idea.

**Comment by ChayimFriedman2** [comment]
> It would be a compiler error, if a more advance type checking was in place. Currently type checking is in-existent at compile time, and is only done at runtime.

Unfortunately, you're right. This is an inherent limitation of single-phase compilers.

**Comment by mhermier** [comment]
Your example is demonstrates why methods should not be shared between 
classes. And mostly make it only valid to use function instead.

**Comment by ChayimFriedman2** [comment]
But functions should have `.call()` as I said, and also suffer from this issue:
```wren
class C {
}
C.f1 = Fn.new {
  this.f2()
}
C.f2 = Fn.new {
}
```

**Comment by ChayimFriedman2** [comment]
I also suspect there are bunch of issue we haven't discovered. It's better to avoid this idea completely, especially since there are excellent alternatives.

**Comment by mhermier** [comment]
No, this is an abuse/dirty hack. You have to consider `this` as the first argument of the function, like what you would do with method introspection, and you are done.

**Comment by ChayimFriedman2** [comment]
And what about `.call()`? And the issue in my example?

**Comment by ChayimFriedman2** [comment]
And where is the dirty hack here?

**Comment by mhermier** [comment]
Example my way, that don't require a grammar hack so this is the real object not the method:
```
class C {
}
C.f1 = Fn.new {|self|
  self.f2()
}
C.f2 = Fn.new {
}
```

**Comment by ChayimFriedman2** [comment]
OK. But I think this is confusing and we all better with the previous ideas. Your thoughts?

**Comment by mhermier** [comment]
My thought is that has I previously stated, this is an expected mess. And your syntax, with `this` in the function, is pretty confusing, because this could also represent the function object itself. Introducing a new syntax for free methods could also be a solution. But each of the solutions seems to have pretty huge cons, mostly because it requires type inference.

**Comment by ChayimFriedman2** [comment]
I'm talking about adding a `reverse` or `reversable` keyword, to indicate reverse operator or operator that can be applied in reverse, accordingly.

**Comment by mhermier** [comment]
Grammatically wise it seems to be the less intrusive, thought it might
requires to adapt a bunch of code to detect the rhs argument type. It also
means that the reverse call is opt-in by the original caller, which can
also be a minor problem in 3rd party code where maintainer is not willing
to be compliant.
Implementation wise it might be a mess. Either you return the error by
return value or by exception, first implementation being a little bit less
worse than the other, it will introduce a conditional jmp (and creation of
fibers in the second case) in assembly on every potential operator call
(same problem with the assign operator proposal). Additionally this would
require a new singleton so it can be used inside the VM or that we
introduce VM error enum and its bindings.

**Comment by mhermier** [comment]
Side note about the new lines: I'm sorry, but I was replying over emails on the phone, and it seems that gmail or github add spurious new lines when adding comments.

**Comment by ChayimFriedman2** [comment]
Do you have better suggestions?

**Comment by mhermier** [comment]
Not really, I'm just saying, before implementation that it will have quite some bytecode cost, even creating a dedicated opcode for it will have possibly high cost.
The thing is that problem you are trying to solve is a common but artificial problem caused by code distribution. Since all the source must be feed, you have full control of what you feed and can correct broken operator/code.

**Comment by ChayimFriedman2** [comment]
But you agree that the current state isn't good.

**Comment by mhermier** [comment]
It is not good or bad. For my use case, I consider that I have access to
all the code, and I can fix any incorrect behaviour already. For your use
case, where you want to use external code, this is a requirement if you
don't want to alter external code. But what ideally I really would like to
avoid is to pay for feature that might be barely used.

**Comment by mhermier** [comment]
I meant nearly used in a hot path. There are change that can be considered
good while heavy, if in the cold path. The problem with that change and the
assign operator changes is that it will end up in a hot path at 99% of use
cases.

**Comment by ChayimFriedman2** [comment]
The biggest problem here is the inconsistency. A beginner can spend a lot of hours trying to understand why `[1] * 3` is `[1, 1, 1]` while `3 * [1]` is an error. This is not what users expect from their language.

**Comment by ChayimFriedman2** [comment]
Unfortunately you're right about the downgrade in the performance. I will think about it a bit more trying to find a solution, but one way or the other I think we should implement it. Performance is the second most important feature of a language, but first and foremost we have the language itself - its semantics, syntax and what users want it to be.

**Comment by ChayimFriedman2** [comment]
I thought about it a little bit more, and I think you're wrong. This **seems** to be on the hot path, but it's definitely not.

See: the only impact of this will be in the case of invalid operator. That is, if we do `3 * 3` the VM will immediately call `Num.*(3, 3)` and everything is fine.

The problem starts where we do `3 * [1]`. Previously that was an error (the `Num.*()` method detects the use of a non-number as the right operand and call `Fiber.abort()`). Now it'll call `List.*([1], 3)`. Clearly, errors are a very cold path. Many languages use this fact: V8, AFAIK re-parse the source code with location information (row and column) where there's need to get a stack trace, and so it can skip this step in the first parsing, saving memory and computing power in the regular case. Take this as an example. Instead of emitting an error immediately, we'll have a delay. But it's definitely worth it.

Also, we will gain a small (or not-so-small) benefit: clearer error messages. Don't look at me like that, error messages are very important! Look at this small example:
```wren
class C {
  construct new(f) {
    _f = f
  }
  f {
    return _f
  }
  *(other) {
    return C.new(f * other.f)
}

var c = C.new(5)
System.print(c * 5) // Cryptic error message: "Num does not implement 'f'". Clarification, please?
```
Of course we can check for non-`C` parameter in `C.*()`, but we won't. This is how users write programs.

Now they'll **have to**:
```wren
class C {
  construct new(f) {
    _f = f
  }
  f {
    return _f
  }
  *(other) {
    if (!(other is C)) {
      return System.notImplemented
    }
    return C.new(f * other.f)
}

var c = C.new(5)
System.print(c * 5) // "Unsupported operand types for '*': 'Num' and 'C'"
```

**Comment by mhermier** [comment]
About your first arguments, it is true that it can be disappointing. But
there only 3 ways to fix it:
- You fix it for yourself.
- You can fix it and you propose it to the maintainer.
- You fix it by using a language feature so you don't have to ask the
maintainer of other code.
The exact example you are giving is an example of what should be in the
second way. But what you are trying to do is to push the 3rd way, by
bringing a language feature.
Each of these way have interests, and 1st and 2nd usually goes in pair
usually you do the 1st and when you are confident you do 2nd.
I agree the 3rd way has various interest. But it is not also a universal
solution, like subclassing/separation...
About the performance degradation, I considered even creating a specific
opcode, it can mitigate the problem by joing call and conditional jump in
one opcode (or even fully integrate it in one opcode). But at the end of
the day it is still a band aid, and just optimise the issue, not really
solve it. Maybe that one can be tested, but I don't really have high hopes
because of the complexity it brings to do the error management, and you
also have to garantee there are no abuse of the error or unexpected
propagation of that error and provoke erroneous behaviour. I don't say it
can be worth the effort, but it has to be tested and validated.

**Comment by mhermier** [comment]
You are right, error management should go in the cold path, but how do you
handle the error in the hot path? You must have to alter the hot path in a
way that it is prepared to receive the error. And this is where lies the
issue. There are some cheaper way to do it, but nothing is cheaper than
having nothing to do. In static language this cost is lifted at compile
time, in dynamic language this cost is resolved at runtime and can become
really expensive.

**Comment by ChayimFriedman2** [comment]
This way already exists. Built-in classes check their operand. For others, the impact is not so great.

**Comment by mhermier** [comment]
For your specific issue, you can do it the same way, without the need to
introduce the heavy machinery required by what you originally proposed.
This machinery is useful but you have to introduce it with care and
awareness.

**Comment by ChayimFriedman2** [comment]
I have no specific issue. I've encountered #746 accidentally. And I think that just like there the user got confused from this behavior, many will too.

**Comment by mhermier** [comment]
A reason way I'm a bit reluctant about the implementation you proposed,
consider the following code:
```
for (mat in matrices) {
  norm_mat.append(mat.norm() * mat)
}
```
Your assumption that it is a slow path is thrown away, and if you have
gazillion of matrices, the cold path is the hot path.
I agree this code is a little bit artificial, but you have to consider it,
in its generality, and the probability that the cold path become hot is
very likely if generalised...

**Comment by ChayimFriedman2** [comment]
Where are the performance hits in your example?

**Comment by mhermier** [comment]
`mat.norm() * mat` is the only operator. I agree that my example might be poorly chosen because of the matrix allocation that might be the slowest part.

But lets consider implementation of reverse operator using exception. To perform it, the following logic has to be done (in pseudo code):
```
{
  fiber = Fiber.new()
  push result of mat.norm() on top of fiber
  push mat on top of fiber
  fiber.try("operator")
  if (!fiber.hasError() || fiber.error != NotImplemented) return fiber.value in current fiber
  swap arguments on current fiber
  call("reverse operator")
}
```
This pseudo code already optimize second call by using current fiber. But the first fiber creation on hot path is almost a requirement, unless you use some heavy hackery on the current fiber.
Now imagine that you have to do this on every binary operator in existence... because you'll never know when a reverse operator will occur...

Even if you can reduce/remove some of the code by doing a dedicated opcode and hackery, there is a bunch of extra logic to be added in the hot path.

**Comment by ChayimFriedman2** [comment]
You don't have to raise an error. You can just return value. Then your code become:
```wren
{
  push result of mat.norm()
  push mat
  call("operator")
  if (stack top == NotImplemented)
  {
    pop
    call("reverse operator")
    if (stack top == NotImplemented) Fiber.abort("Unsupported operand types for '*': 'Mat' and 'Mat'")
  }
}
```
Which is essentially two more CPU instructions for the hot path (`cmp stack_top, NotImplemented` and `jnz after_if`) in x86. Probably ARM will have three because we'll have to calculate the stack top, but I'm not sure. Anyway, this is not a big deal.

**Comment by mhermier** [comment]
Well due to how stack works, there is more stack manipulation than that to do. It is a little bit worse than that, since return will remove the first pushes, you have to dup the stack values and restore the stack to propagate the result in the hot path, in cold path you only need to pop and swap the arguments. These are quite cheap in a single usage, I agree. But consider even a single matrix manipulation or statistics, and also the damage this change does to regular math when cumulated, with all the preamble added to the hot path ???

**Comment by ChayimFriedman2** [comment]
Hacky, maybe, but a single `DUP_2_TOPMOST` bytecode instruction will solve the problem. And as I said, I think it's mandatory: first and foremost the users, and just then the performance.

**Comment by mhermier** [comment]
Test it and give numbers. It will show usability, and will give some reference numbers to prove or not my point. And no, nothing is ever mandatory (other than the foundations required for computation), everything (else) is a trade of. Some of my previous patches were refused while resolving the stack corruption issue for good, with marginal effects on the speed (~15% speed decrease). And they where refused because of these marginal effects even if stability is mandatory.

**Comment by mhermier** [comment]
I gave a quick try at implementing it, and found a new corner case. Consider:
```
class Foo {
  construct new() {}
}
class Bar {
  construct new() {}
  reverse +(foo) { 42 }
}
Foo.new() + Bar.new()
```
In that case method will not be able to be resolved, as such if normal evaluation takes place it will throw an exception. Issue can be mitigated by checking for method presence, but that means additional checks in the hot path.

**Comment by ChayimFriedman2** [comment]
I'm working on implementing it, and have an (almost) working solution that bypass this obstacle too. Hope to send a draft PR for few days.

**Comment by ChayimFriedman2** [comment]
I have not benchmarked yet, but from my calculations I except max. 5-6 CPU instructions on the hot path. That means completely opaque effect.

**Comment by mhermier** [comment]
I made an attempt, because it can be done by adding only one opcode but
requires some refactoring of calling methods.

**Comment by ChayimFriedman2** [comment]
Exactly what I did 😃

**Comment by ChayimFriedman2** [comment]
Ah, I added two. One for calling binary operators and one for returning from them.

**Comment by mhermier** [comment]
You can do it all at once, by ensuring the 2 operators are in sequence (or
passing 2 method index, but it doesn't add much). Since the rest is fixed,
this can really be BINOP_CALL dealing all the mess returning result of top
of the stack.

**Comment by ChayimFriedman2** [comment]
But when will I call the reverse operator?

**Comment by mhermier** [comment]
Since it is inside the VM, the opcode can do anything and you only need to
add 1 to get the reverse operator (require a convention). But it need
internal rework, to avoid code repetition.

**Comment by ChayimFriedman2** [comment]
The symbol is same. But I need to call the reverse operator it when it returns, not when it called.

**Comment by mhermier** [comment]
The symbol (as symbol id) can't be same, otherwise you end up
disallowing/overwriting normal operator.

**Comment by mhermier** [comment]
I think we don't understand each other try to do something, if I can do
something I'll try to, but thing can be messy because of the pile of local
changes I have.

**Comment by ChayimFriedman2** [comment]
The symbol is generated by name. Every `*` method, for example, will have the same symbol across the same module.

**Comment by ChayimFriedman2** [comment]
And whilst foreigns and primitives returns immediately, for user methods we need to wait until the `CODE_RETURN` before we know the return value.

**Comment by ChayimFriedman2** [comment]
I went with a `reversible` keyword. However, after I had a complete implementation I've noticed a glitch. Operators can be commutative, but class cannot overload operator with different meanings. Consider the following code from the standard library:
```wren
class List is Sequence {
  +(other) {
    var result = this[0..-1]
    for (element in other) {
      result.add(element)
    }
    return result
  }
}
```
Now, I want to be able to write both `[1, 2, 3] + (4..6)` and `(1..3) + [4, 5, 6]`. But if I mark the operator as reversible, the second expression gives `[4, 5, 6, 1, 2, 3]`.

Here are the solutions I've thought about. I'd like to hear you feedback:
 - Be able to mark another method as `reverse`. I still don't know if I want to remove `reversible` or leave it (it is useful in many operators, but `reverse` can cover that, although with a bit more code). This is my preferred option.
 - Make binary operators take optional argument `isReverse` which would indicate whether this is a straight or reverse call.
 - A really breaking change, inspired by C#, and partly C++: make operators static methods. Then, they'll receive **two** arguments, `lhs` and `rhs`. I don't like this approach, because it'll: 1) be a big breaking change. 2) require a lot of boilerplate code to see where's the class instance. 3) make operators can't access fields.

**Comment by mhermier** [comment]
Can you show the reversible usage? Because your examples is a bit confusing.
Problem you are seeing is that you are trying to augment an operator,
thinking that it is commutative while it's not...

**Comment by ChayimFriedman2** [comment]
No it's not. But this is still natural for users to do `(1..3) + [4, 5, 6]`. An example of a reversible operator is the `List.*` operator in the standard library (there is also String.* which is very similar):
```wren
class List {
  reversible *(count) {
    if (!(count is Num)) return System.notImplemented
    if (!count.isInteger || count < 0) {
      Fiber.abort("Count must be a non-negative integer.")
    }

    var result = []
    for (i in 0...count) {
      result.addAll(this)
    }
    return result
  }
}
```

**Comment by ChayimFriedman2** [comment]
Updated the title to reflect that.

**Comment by mhermier** [comment]
I wonder if your confusion does not come from the fact you think that a
class operator should be reversible or not. For me one should be able to
provide both. I mean, commutativity depends on both operands, so it might
be desirable to provide something in the regular path and something else in
the reversed path to solve ambiguities.

**Comment by ChayimFriedman2** [comment]
This is exactly what I figured some hours ago, and I ask your advice.

**Comment by mhermier** [comment]
I know it since almost the beginning, and I think it is best to have the 2
definitions. Example:
```
class Foo {
  *(rhs) {...}
  reversed *(rhs) {...}
}
```
That way you control both call path separately. Reversed meaning that the
regular call failed, and it is not an ambiguous situation where the
priority start a guessing game.
There can be some issues calling regular and reversed from each other, but
it can be dealt with an external function till things get sorted.
Now that I think again at it, if the reversed method is missing it will
trigger an exception... So there is also a mess of error handling to
consider/resolve...

**Comment by ChayimFriedman2** [comment]
If so, why did you misled me and proposed a way to mark same operator as reverse? 😠 

I said,

> Just like Python, add reverse operator semantics. Something like a reverse keyword, e.g. reverse *(other). This has the disadvantage of adding a keyword and requiring another method for the same operator (usually it'll just return this * other).

And you responded,

> Another solution to this problem could be to allow to expand classes
outside of primary class definition, _**or granting an operator to be
commutative on some class.**_

**Comment by ChayimFriedman2** [comment]
And believe me, yes it was hard, but I did that 😃

**Comment by ChayimFriedman2** [comment]
And I'd still like to hear your opinion about having both `reversible` and `reverse` keyword.

**Comment by mhermier** [comment]
It is an error in communication. What I meant that the commutative operation could be accessible on some class. I didn't explicited the way to do so. Back then in my head it was *trivial* that granting was an addition, not a modification of the current behavior.

To me the keyword is not yet important, it is still just a marker to test the feature. Ideally you should do the search for previous art in other languages on that feature, justify your choice based around prior art and what is different/related to other languages that explain your choice, and hear about other testimonies. You already talk about python I think, but if you could look in some other languages for reference (if there is any) it would be nice. Example C++ handle the situation by allowing external overloading, which is not really applicable in Wren lang. If a consensus emerge you have your choice in the syntax.

For me this feature is quite a novelty, thought I remember about some formal languages allowing commutativity definition. But most of them are statically typed so there is no prior art I can really think of, that would (in)validate using a particular syntax, so I'm still open for ideas. That said, the keyword approach is quite good, thought it add some crust to name resolution for named call.

**Comment by ChayimFriedman2** [comment]
OK. C++ and C# offer similar solution, by not always passing the receiver as the first argument (either by method being static or by not being method at all).
Python strategy, converted to Wren, will mean adding a `reverse` keyword,
Those are the languages I know that have operator overloading. You're more than welcome to add more. In the meantime, I will change my code from `reversible` keyword to `reverse` keyword (it's not so much work - the VM is almost the same; only the compiler and the core library will need more than trivial changes). I hope to open a PR tonight (I'm UTC+03:00).

**Comment by ChayimFriedman2** [comment]
Do you think we need to make having only the reverse operator an error?

**Comment by mhermier** [comment]
I would say the situation is complex and it has to be seen in action to take proper decisions. Problem with error is that you can intercept ones from lower in the call-stack and misinterpret situation. With return value it is a little bit harder but it can also append. So lets have something to toy with and see how hard we hit the wall.

**Comment by ChayimFriedman2** [comment]
Ahm, I'm researching the area now, and Ruby behave the same as Wren. I can understand why 😢

**Comment by ChayimFriedman2** [comment]
OK. Here's a summary of various programming languages and what they do:
 - Lua:
   - Arithmetic operators (`+`, `-`, `*`, `/`, `%`, `^`, `..` (concatenation)) - if the left operand defines the operator, it wins. Otherwise, we look at the right. Operators can know the order of operands because they're passed by order (like static method, but they took a strange decision and made them instance method - so `self` is not the class type 😄).
   - Comparison operators (`==`, `!=`, `<`, `<=`, `>`, `>=`) - the operands must be of the same type and have the same operator method (I don't understand why both constraints are needed - isn't the second implied by the first?). This way, they eradicate the problem. I can't understand the reason behind the differentiation between arithmetic and comparison operators.
 - C++/C# - disconnect operators from instances, and rely on static typing. That is, to be able to overload the reverse version you're free to reorder the operands (no instance method), and to know where you can't handle a type you list all types you know how to handle (static typing). Wren can do the first, but not the second. A viable solution will be to return `System.notImplemented` in order to indicate unknown type (a combination of C++/C# and Python approaches).
 - Dart, Ruby - do not handle this problem. Overloading reverse operators is not possible (technically yes using monkey-patching (extension methods in Dart), but this is strongly discouraged).
 - Python - provides both straight and reverse operators (`__r<op>__` magic methods) and relies on returning `NotImplemented` to know where to call the reverse operator. This is the best solution I've found, but it hardens commutative operators overloading and encourage code duplication, or at least boilerplate. Consider the following code I wrote from the core library of Wren:
```wren
class List {
  *(count) {
    if (!(count is Num)) return System.notImplemented
    if (count < 0 || !count.isInteger) Fiber.abort("Count must be a non-negative integer.")

    var result = []
    for (i in 0...count) {
      result.addAll(this)
    }
    return result
  }
  reverse *(count) { this * count }
}

class String {
  *(count) {
    if (!(count is Num)) return System.notImplemented
    if (count < 0 || !count.isInteger) Fiber.abort("Count must be a non-negative integer.")

    var result = ""
    for (i in 0...count) {
      result = result + this
    }
    return result
  }
  reverse *(count) { this * count }
}

[1, 2, 3] * "not a num";
```
This code does not work. It enters an infinite recursion. Can you see why? `List.*(_)` returns `System.notImplemented`, and the VM calls `String.reverse *(_)`. It calls `String.*(_)` but here we enter _a new series of binary operator calls_, and so after `String.*(_)`, returns `System.notImplemented` the VM calls `List.reverse *(_)`. And so on. This is exactly what Python does in a similar case.
The solution is to not call the straight operator from the reverse one. But it's easy to forget, and even worse, this is a code duplication. We can obviously extract the code to a method, but then we have boilerplate. Either way we lose.

I think the best will be to have **both `reverse` and `reversible` keywords**. This way, commutative operators can be `reversible` while others can overload the reverse operator. I'm waiting for your feedback.

**Comment by mhermier** [comment]
The Haxe language use the `@:commutative` decorator/meta-tag to specify the property, but it is helped by the fact that the language is statically typed.

**Comment by mhermier** [comment]
I just thought about something trivial that needs to be checked, the operator that are involved as left value evaluation expression should be forbidden and checked by the test suite. So far, I think it should at least imply assignment syntax (`this.foo = bar`) and subscript operators (`foo[bar]`), but in the future it might also mean arithmetic assignment operator (`foo += bar`, `foo -= bar`, ...).

**Comment by ChayimFriedman2** [comment]
I've encountered a problem. Consider:
```wren
class C {
  ..(other) { }
  reverse ..(other) { this...(other) }
}
```
The programmer meant to call the straight `..` operator on `C`, but the lexer emits the token `TOKEN_DOTDOTDOT` and the interprets the expression as `this ... other`. Same problem for calling `...`, the parser will emit `TOKEN_DOTDOTDOT` then `TOKEN_DOT` instead of `TOKEN_DOT` then `TOKEN_DOTDOTDOT` required to correctly interpret the expression.

The problem can be solved by adding a space, but can you understand the following code?
```wren
class C {
  ..(other) { }
  reverse ..(other) { this. ..(other) }
}
```

I thought about allowing (requiring?) parentheses around directly called binary operators:
```wren
class C {
  ..(other) { }
  reverse ..(other) { this.(..)(other) }
}
```
This is better, but I still have few questions:
 - Do we need to **allow** them (but also allow separating using space) or **require** them around `..` and `...`, or maybe even require them around any operator?
 - Should we allow them around ordinary named methods (`this.(m)()`) to keep consistency, or not?
 - Maybe we should choose a completely different solution?

I'd like to hear your opinion.

**Comment by ChayimFriedman2** [comment]
And if we allow (or require) parens around every operator, what about reverse ones? Should it be `this.reverse (+)()` or `this.(reverse +)()?

**Comment by mhermier** [comment]
I didn't considered the .. and ... operators in my proposition. But do we
really need to allow them to be reversed? Since the original definition is
pretty one sided (non reversible) I would not consider it a candidate to
the reverse operator.
If we really need to allow them, the space disambiguation is the simplest
solution and it cost absolutely nothing.
Personally I would be against using parenthesis there, I would reserve it
to method evaluation, to separate it from the usual name resolution.

**Comment by ChayimFriedman2** [comment]
I don't see a reason why **not** allowing reverse them. They are binary operators just like others. If anything, I wonder why to allow overloading them at all (Python does not allow overloading the colon operator used to create slices, for example. C# also does not allow it).

The problem with the space is that it's easily slipping out of sight. It's hard to distinguish between `a. ..b` and `a...b`. The only visual clue is the parens, because although `a...(b)` is valid, it's not common. But, for example, `a...(b + 1)` can be written (although `a...b + 1` is the same, many people will prefer to be explicit, especially in complex expressions).

**Comment by ChayimFriedman2** [comment]
And (curiously): what's method evaluation? Do you mean dynamically choosing the name of the method (like `object[property]` in js)?

**Comment by mhermier** [comment]
The parenthesis of parameters option is bad, it complexify the compiler,
that needs to elaborate on the argument. I  addition, there can be some
legitimate case where parenthesis are required.

**Comment by mhermier** [comment]
Something more like c++ operator ->* where rhs is a computer method object.

**Comment by ChayimFriedman2** [comment]
But Wren does not allow you to read method object 😉

**Comment by ChayimFriedman2** [comment]
And can you give an example of case where parentheses are required please?

**Comment by ChayimFriedman2** [comment]
Why is it complicates the compiler? I mean, it does complicate things, but not so much.

**Comment by mhermier** [comment]
It doesn't allow for now, but we should consider it as a possibility, a
don't shoot ourself in the foot by not looking in the future...
I don't have all the operator precedence in mind, but there is no guarantee
that the .. operator will have the smallest precedence, there will always
be possible ambiguous case were parenthesis could be required.
It complicated the compiler because technically while it can still be done
with an LL(1) language, I think it would make it an LL(2) or even LL(3).

**Comment by ChayimFriedman2** [comment]
No there are not such places. We're talking about wrapping **the operator itself** (`a(..)b`), not **the operator with its operands** (`(a..b)`).

**Comment by mhermier** [comment]
I see but how would you implement method call (without mirroring since it
would fall in the `call` protocol)?

**Comment by ChayimFriedman2** [comment]
I don't understand the problem, sorry.

**Comment by ChayimFriedman2** [comment]
If you mean `a(..)b` is ambiguous with `a(123)`, then you're right. This was just an example, but since we're talking about operator invocation there will be always a dot before: `a.(...)b`.

**Comment by mhermier** [comment]
Given the fact that you could get/store a method object into a variable,
give a syntax to call that stored method on a given object.

**Comment by ChayimFriedman2** [comment]
But where is the problem here?

**Comment by mhermier** [comment]
There is no real problem. I just find it not elegant at all, and gives ()
an usual meaning of special name resolution. I  contrary, I find
`obj.(method)(params)` (where all names are variables) more elegant, even
if it means that for 2 operators, it requires a space to disambiguate the
call of `obj. ..(param)`.

**Comment by ChayimFriedman2** [comment]
OK.

**Comment by ChayimFriedman2** [comment]
Got it. Although we can (with LL(2) parser) have both.

**Comment by mhermier** [comment]
Well there is always some possible ambiguity somewhere that I didn't think
off. But the more I think of the `a(..)b syntax the more ugly I found it

**Comment by ChayimFriedman2** [comment]
I personally don't agree. This is what F# has, and it's perfect. But I'll ignore my opinion 😉

**Comment by mhermier** [comment]
You can "not agree" with me, in the contrary it is good. But bring some
counter arguments/solutions to the table. It is not because one language
does it, that it is an absolute solution. They could have been trapped by
the grammar, been lacking of imagination, been referencing an obscure
solution taken from what ever language....

**Comment by ChayimFriedman2** [comment]
F# has a single-phase compiler, two. Already one similarity 😄.
In fact, I brought solutions and you rejected them 😃.
I am not one of F# developers, and can only guess the reasons. But it seems reasonable that the limitations of single-phase compilers enforced that. 
But we're both agree that both solutions are possible technically. All our controversy is which one is more readable.
And finally, as I already said, I accepted your opinion for the implementation. We're arguing, ahm, in the air.

**Comment by mhermier** [comment]
I don't rejected them, I don't find it viable if you consider method
invocation.
If you have a solution to express method invocation, operator or what ever
you are welcome...

**Comment by mhermier** [comment]
I think that the functionality to direct call function should be moved to
this own thread/pull request. I mean, even if I like the dot syntax to
extends, it is its own can of worms than what the original proposition was
about.

**Comment by ChayimFriedman2** [comment]
I don't agree since the only valid use case for this is for reverse operators. I even considered making other uses an error.
